### PR TITLE
Add doCheckDisplayNameOrNull to jenkins core

### PR DESCRIPTION
**Pull request by krisstern**
See [JENKINS-69916](https://issues.jenkins.io/browse/JENKINS-69916).

### Testing done

Double checked no existing test has been caused to fail by runnig `mvn verfiy`. No test has been added since this will be done downstream. 

### Proposed changelog entries

- Move `doCheckDisplayNameOrNull` from `AbstractProject` to `TopLevelItemDescriptor` to allow reuse in pipeline.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Comment by daniel-beck** [comment]
This needs an explanation, because the code location is unexpected.

File: core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java
Comment (jglick) [inline core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:28]: What does this new method have to do with detached plugins?

Comment (krisstern) [reply core/src/main/java/jenkins/plugins/DetachedPluginsUtil.java:28]: It doesn't, I wasn't sure where to place the code at first actually

**Comment by jglick** [comment]
Does not seem to align with

> an AbstractItemDescriptor interface with a default implementation for that doCheckDisplayNameOrNull

by @mawinter69 in https://github.com/jenkinsci/workflow-job-plugin/pull/411#issuecomment-1925271400

**Comment by mawinter69** [comment]
I would expect this to be in `TopLevelItemDescriptor` or a dedicated interface that plugins implement, `Job` is not sufficient as we also need to check for duplicates with Folders.

**Comment by krisstern** [comment]
> This needs an explanation, because the code location is unexpected.
> 
> Why did you choose to do this rather than pull up
> 
> https://github.com/jenkinsci/jenkins/blob/b0baebaf880047ac32428fd58abf62a20e6f9a54/core/src/main/java/hudson/model/AbstractProject.java#L1943-L1946
> 
> into `Job`? Seems like that would automatically take care of the plugin.
> > [workflow-job] Un-inlining WorkflowJob/configure-entries.jelly
> 
> Would you expect an average Jenkins admin to be able to understand this change and how it impacts them?

Sure, I just moved the code to `Job`. Actually I wasn't sure where to place the code originally, so I was making a wild guess. For more context please see https://github.com/jenkinsci/workflow-job-plugin/pull/411.

**Comment by krisstern** [comment]
> TopLevelItemDescriptor

No problem, let me move it again.

File: core/src/main/java/hudson/model/Job.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Job.java:56]: Revert, we're not doing star imports in this repo.

File: core/src/main/java/hudson/model/TopLevelItemDescriptor.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/TopLevelItemDescriptor.java:43]: As above.

File: core/src/main/java/hudson/model/TopLevelItemDescriptor.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/TopLevelItemDescriptor.java:291]: This cannot have been tested, because `WorkflowJob` is not an `AbstractProject`. Something like the following would be needed here (plus fix imports as needed),

```suggestion
    public static FormValidation doCheckDisplayNameOrNull(@AncestorInPath TopLevelItem<?, ?> item, @QueryParameter String value) {
        return Jenkins.get().doCheckDisplayName(value, item.getName());
```

File: core/src/main/java/hudson/model/TopLevelItemDescriptor.java
Comment (mawinter69) [inline core/src/main/java/hudson/model/TopLevelItemDescriptor.java:290]: You can't use AbstractProject here. This must be TopLevelItem
`@AncestorInPath TopLevelItem item`

File: core/src/main/java/hudson/model/Job.java
Comment (krisstern) [reply core/src/main/java/hudson/model/Job.java]: Reverted

File: core/src/main/java/hudson/model/TopLevelItemDescriptor.java
Comment (krisstern) [reply core/src/main/java/hudson/model/TopLevelItemDescriptor.java:43]: Reverted

**Comment by mawinter69** [comment]
created https://issues.jenkins.io/browse/JENKINS-72988 for the wrong check

File: core/src/main/java/hudson/model/Job.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Job.java:131]: Please revert. While technically correct, would prefer no unrelated cleanup in places not otherwise touched in the PR.

**Comment by daniel-beck** [comment]
Looks reasonable now. Please file a downstream PR once the incremental is deployed, demonstrating this works as expected, then check

> For new APIs and extension points, there is a link to at least one consumer.

**Review by mawinter69 — CHANGES_REQUESTED** [review]
We should remove the implementation from `AbstractProject` as it is obsolete now

**Comment by mawinter69** [comment]
> Looks reasonable now. Please file a downstream PR once the incremental is deployed, demonstrating this works as expected, then check
> 
> > For new APIs and extension points, there is a link to at least one consumer.

The workflow-job plugin should now work properly without any change as it didn't yet implement the method, we would need PRs in folder and matrix project plugins that remove the implementation there

File: core/src/main/java/hudson/model/Job.java
Comment (krisstern) [reply core/src/main/java/hudson/model/Job.java]: Done

**Comment by krisstern** [comment]
> We should remove the implementation from `AbstractProject` as it is obsolete now

Done

**Review by mawinter69 — APPROVED** [review]

**Comment by daniel-beck** [comment]
Still needs downstream PRs.

Also, the proposed changelog entry needs work:

> [workflow-job] Un-inlining WorkflowJob/configure-entries.jelly

Changelog audience is generally Jenkins users and administrators who we need to assume know nothing about Jenkins development.

**Comment by krisstern** [comment]
Updated the "proposed changelog entries" in the description above

**Comment by krisstern** [comment]
> Still needs downstream PRs.
> 
> Also, the proposed changelog entry needs work:
> 
> > [workflow-job] Un-inlining WorkflowJob/configure-entries.jelly
> 
> Changelog audience is generally Jenkins users and administrators who we need to assume know nothing about Jenkins development.

Downstream changes made in https://github.com/jenkinsci/workflow-job-plugin/pull/411.

**Review by daniel-beck — APPROVED** [review]
Works great with the PR build and the downstream PR build.

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
